dgraph/cmd/graphql/handler.go

```diff
@@ -0,0 +1,171 @@
+//
+// - If an error was encountered before execution begins,
+//   the data entry should not be present in the result.
+//
+// - If an error was encountered during the execution that
+//   prevented a valid response, the data entry in the response should be null.
+//
+// - If no errors were encountered during the requested operation,
+//   the errors entry should not be present in the result.
+//
+// - There's rules around how errors work when there's ! fields in the schema
+//   https://graphql.github.io/graphql-spec/June2018/#sec-Errors-and-Non-Nullability
+//
+// - The "message" in an error is required, the rest is up to the implementation
+//
+// - The "data" works just like a Dgraph query
+//
+
+type graphqlHandler struct {
+	dgraphClient *dgo.Dgraph
+	schema       *ast.Schema
+}
+
+type graphqlRequest struct {
+	Query         string                 `json:"query"`
+	OperationName string                 `json:"operationName"`
+	Variables     map[string]interface{} `json:"variables"`
+}
+
+func (gh *graphqlHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+
+	var gqlReq graphqlRequest
+
+	switch r.Method {
+	case http.MethodGet:
+		// fill gqlReq in
+	case http.MethodPost:
+		mediaType, _, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+		if err != nil {
+			// error response - no data
+			return
+		}
+
+		switch mediaType {
+		case "application/json":
+			if err = json.NewDecoder(r.Body).Decode(&gqlReq); err != nil {
+				// error response - no data
+				return
+			}
+		default:
+			// nothing else is valid?
+			//
+			// error response - no data
+			return
+		}
+	default:
+		// error response - no data
+		return
+	}
+
+	doc, gqlErr := parser.ParseQuery(&ast.Source{Input: gqlReq.Query})
+	if gqlErr != nil {
+		// error response - no data
+		return
+	}
+
+	listErr := validator.Validate(gh.schema, doc)
+	if len(listErr) != 0 {
+		// error response - no data
+		return
+	}
+
+	op := doc.Operations.ForName(gqlReq.OperationName)
+	if op == nil {
+		// error response - no data
+		return
+	}
+
+	// actually need the output vars here because there's been some type magic
+	// done on them in the validator
+	_, err := validator.VariableValues(gh.schema, op, gqlReq.Variables)
+	if err != nil {
+		// error response - no data
+		return
+	}
+
+	switch op.Operation {
+	case ast.Query:
+		resp := &graphQLResponse{
+			Data: []byte(`{ "lifeQuery": [ { "meaning" : 42 } ] }`),
+		}
+
+		b, err := json.Marshal(resp)
+		if err != nil {
+			// error response - "data": null
+			return
+		}
+		_, err = w.Write(b)
+		if err != nil {
+			// nothing to do, just log?
+		}
+	case ast.Mutation:
+		b, err := json.Marshal(ErrorResponse("Not yet implemented"))
+		if err != nil {
+			// error response - "data": null
+			return
+		}
+		_, err = w.Write(b)
+		if err != nil {
+			// nothing to do, just log?
+		}
+	default:
+		// error response - no data
+		return
+	}
+}
```

Comment:
This is a really long function, should break into multiple functions.

Reply:
Thank!  I've decided I'm no so worried about the long functions at them moment cause there's lots more to go in there and they'll get broken up as part of that.